### PR TITLE
is137 Add settings for operating system specific paths

### LINKED ISSUE #137: Add system specific vscode-neovim.neovimPath settings entries
I use VSCode 'Settings Sync' extension and I have a personal laptop with Linux and a work laptop with Windows. I use vscode-neovim on both. 

Lately I uploaded my settings to Settings Sync from the Linux laptop with the `vscode-neovim.neovimPath` set to `/usr/bin/nvim`. Then I moved to my Windows work laptop - my neovim path was `C:\tools\nvim\bin\nvim.exe` and vscode-neovim was working ok. But then I downloaded the fresh settings and the `neovimPath` on Windows changed to the Linux one - `/usr/bin/nvim` and vscode-neovim stopped working.

Perhaps it would be cool to split the neovimPath into three separate settings:
`neovimPath.windows`,
`neovimPath.linux`,
`neovimPath.osx`,

VSCode already uses the convention I proposed with their `terminal.external.[linux/windows/osx]Exec` setting
obraz

### PR DESCRIPTION
**Pull request by 3nuc**
Fixes #137 

This branch adds operating system specific paths (neovimPath). ~I still need to add some tests~ and actually manually check if it even works.

Anyways, I'm working on it. Uploading a draft PR as a heads up

- Create settings and logic for overriding settings on Windows, Linux and OSX:
  - [X] neovimPath
  - [X] neovimInitPath (logic is written but I didn't replace the logic in `extension.ts` yet
- [X] Update package.json setting descriptions
- [X] Mark neovimPath and neovimInitPath as deprecated in settings.json
- [ ] ~**Figure out how to mock `vscode.workspace.getConfiguration('vscode-nevoim')` in unit tests so I can write unit tests for that functionality**~

neovimPath and neovimInitPath logic is written here in such a way that if neovimPath exists, the system specific settings are ignored. So people who are oblivous to the deprecation of neovimPath won't have a broken workflow when this is merged.

### CODE REVIEW

**Comment by asvetliakov** [comment]
Thank you for the PR! Don't have much time recently, but will try to review in short time

**Comment by 3nuc** [comment]
About the testing, as a note to self:
I guess I could mock `vscode.workspace.getConfiguration` so it's a class that has a getter `get` and an internal property `settings` whose properties could be
```json
{
"some.long.dotted.stuff": 123
}
```

File: src/utils.ts [hunk 1]
```diff
@@ -1,5 +1,8 @@
+import { workspace, TextEditor, TextDocumentContentChangeEvent, Position } from "vscode";
+
+import { EXT_NAME } from "./extension";
```
Comment (asvetliakov) [inline src/utils.ts]: This may cause a circular reference issues. I'd suggest to move EXT_NAME/EXT_ID to utils

File: src/utils.ts [hunk 2]
```diff
@@ -333,3 +336,49 @@ export function isCursorChange(change: TextDocumentContentChangeEvent, cursor: P
...
+    settingPrefix: settingPrefix,
+    legacySetting: { environmentVariableName?: "NEOVIM_PATH"; vscodeSettingName: legacySettingName },
+): string | undefined {
+    const settings = workspace.getConfiguration(EXT_NAME);
+    const isWindowsSubsystemForLinux = settings.get("useWSL");
+
+    //https://github.com/microsoft/vscode/blob/master/src/vs/base/common/platform.ts#L63
+    const platform = process.platform as "win32" | "darwin" | "linux";
+    const mapPlatformToSetting: { [platformName in typeof platform]: string } = {
+        win32: `${settingPrefix}.windows`,
```
Comment (3nuc) [inline src/utils.ts]: I should probably drop the map here and have win32: 'settingsPrefix.win32'

File: src/utils.ts [hunk 3]
```diff
@@ -333,3 +336,49 @@ export function isCursorChange(change: TextDocumentContentChangeEvent, cursor: P
     }
     return false;
 }
+
+export type legacySettingName = "neovimPath" | "neovimInitPath";
```
Comment (asvetliakov) [inline src/utils.ts]: Types should be upper-case

File: package.json [hunk 4]
```diff
@@ -42,12 +42,48 @@
...
+					"title": "Neovim path on Windows",
+					"description": "Full path to Neovim executable that should be used if using Code on Windows. If the neovimPath setting is not empty exists it will be used instead of this setting"
+				},
+				"vscode-neovim.neovimOperatingSystemPaths.osx": {
+					"type": "string",
+					"default": "",
+					"title": "Neovim path on OSX",
+					"description": "Full path to Neovim executable that should be used if using Code on OSX. If the neovimPath setting is not empty it will be used instead of this setting"
+				},
+				"vscode-neovim.neovimOperatingSystemPaths.linux": {
```
Comment (3nuc) [inline package.json]: It should just be `neovimPath.windows` instead of the super lenghty name. Dunno why I changed it

File: package.json [hunk 5]
```diff
@@ -42,12 +42,48 @@
...
 					"type": "string",
 					"default": "",
 					"title": "Custom init.vim path",
 					"description": "Full path to custom neovim init file, equals to startup option -u. If checked useWSL flag specify a Linux path. You can also use exists('g:vscode') in your init.vim to check if neovim is being run in vscode and share init file between neovim/vscode"
 				},
+				"vscode-neovim.neovimInitOperatingSystemPaths.windows": {
+					"type": "string",
+					"default": "",
+					"title": "Custom init.vim path",
+					"description": "Full path to custom neovim init file, equals to startup option -u. If checked useWSL flag specify a Linux path. You can also use exists('g:vscode') in your init.vim to check if neovim is being run in vscode and share init file between neovim/vscode"
```
Comment (3nuc) [inline package.json]: Need to update descriptions

**Comment by asvetliakov** [comment]
> About the testing, as a note to self:
> I guess I could mock `vscode.workspace.getConfiguration` so it's a class that has a getter `get` and an internal property `settings` whose properties could be
> 
> ```json
> {
> "some.long.dotted.stuff": 123
> }
> ```
> 
> then the `get` would return properties from `settings`

Just don't bother with testing settings. Small value, big headache

**Comment by 3nuc** [comment]
Went to test out if the extension successfully finds files when using:

- WSL - nvim and custom init.vim successfuly loaded into extension
- Regular Windows - nvim and custom init.vim successfuly loaded into extension
- Regular Linux - nvim and custom init.vim successfuly loaded into extension

I only didn't check OSX (darwin) but should be fine

**Comment by 3nuc** [comment]
Agh didn't consider the edge case where someone has useWSL enabled but the user platform is OSX. It will still try to use the linux path. The goal is to have settings behave in such a way that someone has 3 computers with windows, linux and osx and it still works

File: src/utils.ts [hunk 6]
```diff
@@ -333,3 +337,46 @@ export function isCursorChange(change: TextDocumentContentChangeEvent, cursor: P
...
+    const platform = process.platform as "win32" | "darwin" | "linux";
+
+    const legacyEnvironmentVariable =
+        legacySetting.environmentVariableName && process.env[legacySetting.environmentVariableName];
+
+    //some system specific settings can be loaded from process.env and value from env will override setting value
+    const legacySettingValue = legacyEnvironmentVariable || settings.get(legacySetting.vscodeSettingName);
+    if (legacySettingValue) {
+        return legacySettingValue;
+    } else if (isUseWindowsSubsystemForLinux && platform !== "darwin") {
```
Comment (asvetliakov) [inline src/utils.ts]: I think there should be platform === "win32" only since you can't use WSL on native linux 😄

File: src/utils.ts [hunk 7]
```diff
@@ -333,3 +337,46 @@ export function isCursorChange(change: TextDocumentContentChangeEvent, cursor: P
...
+    } else {
+        return settings.get(`${settingPrefix}.${platform}`);
+    }
+}
+
+export function getNeovimPath(): string | undefined {
+    const legacySettingInfo = {
+        vscodeSettingName: "neovimPath",
+        environmentVariableName: "NEOVIM_PATH",
+    } as Parameters<typeof getSystemSpecificSetting>[1];
```
Comment (asvetliakov) [inline src/utils.ts]: Why so difficult ? just 
```ts
return getSystemSpecificSettings("neovimExecutablePaths", { 
  vscodeSettingsName: "neovimPath", 
  environmentVariableName: "NEOVIM_PATH" 
});
```

File: src/utils.ts [hunk 8]
```diff
@@ -333,3 +337,46 @@ export function isCursorChange(change: TextDocumentContentChangeEvent, cursor: P
...
+    return getSystemSpecificSetting("neovimExecutablePaths", legacySettingInfo);
+}
+
+export function getNeovimInitPath(): string | undefined {
+    const legacySettingInfo = {
+        vscodeSettingName: "neovimInitPath",
+    } as Parameters<typeof getSystemSpecificSetting>[1];
+    return getSystemSpecificSetting("neovimInitVimPaths", legacySettingInfo);
```
Comment (asvetliakov) [inline src/utils.ts]: Same as above

File: src/utils.ts (hunk 6 above)
Comment (3nuc) [reply src/utils.ts]: Fixed. I'm not very smart you know lol

**Comment by 3nuc** [comment]
All done

**Comment by asvetliakov** [comment]
Thank for the PR!

**Pull request merged**